heat/core/random.py

```diff
@@ -1,44 +1,376 @@
+    size = comm.Get_size()
+    max_count = 0xffffffff if dtype == torch.int32 else 0xffffffffffffffff
+    # extract the counter state of the random number generator
+    if dtype is torch.int32:
+        c_0 = (__counter & (max_count << 32)) >> 32
+        c_1 = __counter & max_count
+    else:  # torch.int64
+        c_0 = (__counter & (max_count << 64)) >> 64
+        c_1 = __counter & max_count
+
+    total_elements = np.prod(shape)
+    if total_elements > 2 * max_count:
+        raise ValueError('Shape is to big with {} elements'.format(total_elements))
+
+    if split is None:
+        values = total_elements / 2
+        even_end = values % 2 == 0
+        lslice = slice(None) if even_end else slice(None, -1)
+        start = c_1
+        end = start + int(values)
+        lshape = shape
+    else:
+        offset, lshape, _ = comm.chunk(shape, split)
+        counts, displs, _ = comm.counts_displs_shape(shape, split)
+
+        # Calculate number of local elements per process
+        local_elements = [total_elements / shape[split] * counts[i] for i in range(size)]
+        cum_elements = np.cumsum(local_elements)
+
+        # Calculate the correct borders and slices
+        even_start = True if rank == 0 else cum_elements[rank-1] % 2 == 0
+        start = c_1 if rank == 0 else int(cum_elements[rank-1] / 2) + c_1
+        elements = local_elements[rank] / 2
+        lslice = slice(None)
+        if even_start:
+            # No overlap with previous processes
+            if elements == int(elements):
+                # Even number of elements
+                end = int(elements)
+            else:
+                # Odd number of elements
+                end = int(elements) + 1
+                lslice = slice(None, -1)
+        else:
+            # Overlap with previous processes
+            if elements == int(elements):
+                # Even number of elements
+                end = int(elements) + 1
+                lslice = slice(1, -1)
+            else:
+                # Odd number of elements
+                end = int(elements) + 1
+                lslice = slice(1, None)
+        start = int(start)
+        end += start
+
+    # Check x_1 for overflow
+    lrange = [start, end]
+    signed_mask = 0x7fffffff if dtype == torch.int32 else 0x7fffffffffffffff
+    diff = 0 if lrange[1] <= signed_mask else lrange[1] - signed_mask
+    lrange[0], lrange[1] = lrange[0] - diff, lrange[1] - diff
+
+    # create x_1 counter sequence
+    x_1 = torch.arange(*lrange, dtype=dtype)
+    while diff > signed_mask:
+        # signed_mask is maximum that can be added at a time because torch does not support unit64 or unit32
+        x_1 += signed_mask
+        diff -= signed_mask
+    x_1 += diff
+
+    # generate the x_0 counter sequence
+    x_0 = torch.empty_like(x_1)
+    diff = c_0 - signed_mask
+    if diff > 0:
+        # same problem as for x_1 with the overflow
+        x_0.fill_(signed_mask)
+        while diff > signed_mask:
```

Comment:
This is somewhat sub-optimal from a performance perspective. Would be better to have something like vectorized version of this without a while-loop. Potentially determine the split point and then do the minus and plus magic?

Reply:
With the check `total_elements <= 2 * 0xffffffff` this loop runs for a maximum of four times. However, adding a value larger than 0x7fffffff is not possible and this iterative process also takes care of to correct overflow after 0xffffffff is reached. I can not think of a direct way to get this expected overflow because at some point the sign needs to be flipped.

Reply:
Yes, sorry you explained that obviously already. Nevermind